File: Medium/1026. Maximum Difference Between Node and Ancestor/Solution.py
class Solution(object):
    def __init__(self):
        self.ans = 0

    def update(self, root, val):
        if root == None: return
        self.ans = max(self.ans, abs(root.val - val))
        self.update(root.left, val)
        self.update(root.right, val)

    def dfs(self, root):
        if root == None: return
        self.update(root, root.val)
        self.dfs(root.left)
        self.dfs(root.right)

    def maxAncestorDiff(self, root):
        self.ans = 0
        self.dfs(root)
        return self.ans

File: Medium/1026. Maximum Difference Between Node and Ancestor/Solution.js
let ans;

let update = (root, val) => {
    if(root === null) return;
    ans = Math.max(ans, Math.abs(root.val - val));
    update(root.left, val);
    update(root.right, val);
};

let dfs = (root) => {
    if(root === null) return;
    update(root, root.val);
    dfs(root.left);
    dfs(root.right);
};

var maxAncestorDiff = function (root) {
    ans = 0;
    dfs(root);
    return ans;
};

